src/Build/BackEnd/Components/RequestBuilder/IntrinsicTasks/ItemGroupLoggingHelper.cs

```diff
@@ -95,53 +101,76 @@ internal static string GetParameterText(string prefix, string parameterName, ILi
         /// </summary>
         internal static string GetStringFromParameterValue(object parameterValue)
         {
-            var type = parameterValue.GetType();
-
-            ErrorUtilities.VerifyThrow(!type.IsArray, "scalars only");
-
-            if (type == typeof(string))
+            using (var sb = new ReuseableStringBuilder())
             {
-                return (string)parameterValue;
+                AppendStringFromParameterValue(sb, parameterValue);
+                return sb.ToString();
             }
-            else if (type.GetTypeInfo().IsValueType)
+        }
+
+        // Avoid allocating a temporary list to hold metadata for sorting every time.
+        // Each thread gets its own copy.
+        [ThreadStatic]
+        private static List<KeyValuePair<string, string>> keyValuePairList;
+
+        private static void AppendStringFromParameterValue(ReuseableStringBuilder sb, object parameterValue)
+        {
+            if (parameterValue is string text)
             {
-                return (string)Convert.ChangeType(parameterValue, typeof(string), CultureInfo.CurrentCulture);
+                sb.Append(text);
             }
-            else if (typeof(ITaskItem).GetTypeInfo().IsAssignableFrom(type.GetTypeInfo()))
+            else if (parameterValue is ITaskItem item)
             {
-                var item = ((ITaskItem)parameterValue);
-                string result = item.ItemSpec;
+                sb.Append(item.ItemSpec);
 
                 var customMetadata = item.CloneCustomMetadata();
+                int count = customMetadata.Count;
 
-                if (customMetadata.Count > 0)
+                if (count > 0)
                 {
-                    result += "\n";
-                    var names = new List<string>();
+                    sb.Append('\n');
 
-                    foreach (string name in customMetadata.Keys)
+                    // need to initialize the thread static on each new thread
+                    if (keyValuePairList == null)
                     {
-                        names.Add(name);
+                        keyValuePairList = new List<KeyValuePair<string, string>>(count);
```

Comment:
There's not a way to ensure `count` capacity in the already-allocated case, right? Because setting `Capacity` could result in reallocating to smaller, which isn't what we want.

Reply:
This list will always grow and never shrink. So once it grows to a sufficient size it stops rejiggering altogether. Clear() at the end keeps the Capacity unchanged.

Reply:
There's an EnsureCapacity() on List<T> which is exactly what we want, but it's internal. No biggie, I think it will only resize the first few times and that's it.

Reply:
Actually we could do if (Capacity < count) Capacity = count; but I'd rather rely on List's internal doubling algorithm.

Reply:
Just for clarity, since this is under `if (keyValuePairList == null)`, it will only look at count once per thread, so it will resize beyond when one gets to maximum capacity, but it shouldn't get smaller.